mlrun/artifacts/base.py

```diff
@@ -324,27 +323,120 @@ def base_dict(self):
                 struct[field] = val.base_dict()
         return struct
 
-    def upload(self):
-        """internal, upload to target store"""
+    def upload(self, artifact_path: str = None):
+        """
+        internal, upload to target store
+        :param artifact_path: required only for when generating target_path from artifact hash
+        """
         src_path = self.spec.src_path
         body = self.get_body()
         if body:
-            self._upload_body(body)
+            self._upload_body(body=body, artifact_path=artifact_path)
         else:
             if src_path and os.path.isfile(src_path):
-                self._upload_file(src_path)
+                self._upload_file(source_path=src_path, artifact_path=artifact_path)
+
+    def _upload_body(self, body, target=None, artifact_path: str = None):
+        body_hash = None
+        if not target and not self.spec.target_path:
+            if not mlrun.mlconf.should_generate_target_path_from_artifact_hash():
+                raise mlrun.errors.MLRunInvalidArgumentError(
+                    "Unable to resolve target path, no target path is defined and "
+                    "mlrun.mlconf.artifacts.generate_target_path_from_artifact_hash is set to false"
+                )
+            body_hash, self.spec.target_path = self.resolve_body_target_hash_path(
+                body, artifact_path
+            )
 
-    def _upload_body(self, body, target=None):
-        if calc_hash:
-            self.metadata.hash = blob_hash(body)
+        self.metadata.hash = self.metadata.hash or body_hash
+        if mlrun.mlconf.should_calculate_artifact_hash():
+            self.metadata.hash = calculate_blob_hash(body)
```

Comment:
Did you forget this?

Reply:
I changed the condition instead

Reply:
I think this is not good. What if the artifact already had a hash and we modified the body? In this case, the line `self.metadata.hash = self.metadata.hash or body_hash` would keep the old hash rather than update it. 
My suggestion doesn't have this problem - it will update `self.metadata.hash` anyway, but will not re-calculate.